spatialprofilingtoolbox/ondemand/providers/pending_provider.py

```diff
@@ -0,0 +1,194 @@
+"""Abstract class for providers that need to wait for on demand calculations to complete."""
+
+from abc import ABC, abstractmethod
+from threading import Thread
+from datetime import datetime
+
+from pandas import DataFrame
+
+from spatialprofilingtoolbox.db.database_connection import DBCursor
+from spatialprofilingtoolbox.db.exchange_data_formats.metrics import PhenotypeCriteria
+from spatialprofilingtoolbox.ondemand.providers import Provider
+from spatialprofilingtoolbox.standalone_utilities.log_formats import colorized_logger
+
+logger = colorized_logger(__name__)
+
+
+class PendingProvider(Provider, ABC):
+    """Provide precalculated metrics or start, wait, receive, and then provide them."""
+
+    def get_metrics(
+        self,
+        study_name: str,
+        **kwargs: int | PhenotypeCriteria | list[PhenotypeCriteria]
```

Comment:
Amazing, you can type hint arbitrary anonymous keyword arguments? I did not know.